bootstrapWorktree: add npm install step + optional `--build-all` for SDK/test-ready worktrees

## Context

Per @tobiu's session-end direction (2026-04-26): worktrees created by Claude Code (`.claude/worktrees/<name>/`) start without `node_modules` and without prerequisite build artifacts (e.g., `dist/parse5.mjs`). `ai/scripts/bootstrapWorktree.mjs` already handles the gitignored `ai/mcp/server/*/config.mjs` copies and the optional `.neo-ai-data/` symlink, but stops short of making the worktree truly "test-ready" or "SDK-consuming-script-ready" out of the box.

## The Problem

When picking up a ticket in a fresh worktree that requires running unit tests or invoking SDK-consuming scripts, agents currently hit a multi-step manual bootstrap:

1. `node ai/scripts/bootstrapWorktree.mjs --link-data` — configs + data symlink
2. `npm install` — node_modules
3. `npm run bundle-parse5` — build prerequisite for unit-test config (not always obvious it's needed)
4. *(for frontend work)* `npm run build-all` — bundled dist artifacts

Steps 2-4 are not codified anywhere agents can discover them. Empirically observed during #10339 implementation: ticket-intake passed, branch-before-code created, then test runner failed with `Cannot find module '.../dist/parse5.mjs'`. Each unfamiliar agent re-derives the chain.

## The Architectural Reality

`bootstrapWorktree.mjs` is the canonical entry point per `AGENTS_STARTUP.md §5 Worktree Bootstrap (Claude Code)`. Adding npm-install + optional build steps there means:

- Agents already know to invoke it (it's in the boot mandate).
- The script's existing flag-based design (`--link-data`, `--force`) accommodates new opt-in flags cleanly.
- No new entry points to document; no new mental model.

The trade-off is install-time cost: `npm install` takes ~17s on this machine (808 packages installed against an existing local cache). For frontend workflows, `npm run build-all` is heavier (~minutes). Hence opt-in flags rather than always-on.

## The Fix

Extend `ai/scripts/bootstrapWorktree.mjs` with two new opt-in flags:

1. **`--install`** — runs `npm install` if `node_modules/` does not already exist in the worktree. Idempotent: skip when present. Empirical anchor: ~17s for the install path observed in this session.

2. **`--build-all`** — runs `npm run build-all` (or `npm run bundle-parse5` as a lighter subset) for frontend-affecting tickets. Implies `--install`. Idempotent: skip if `dist/parse5.mjs` already exists AND the worktree is up-to-date with origin/dev (timestamp check).

Recommended composition: agents pick up backend/MCP tickets with `--install`; agents pick up frontend tickets with `--build-all`. Update `AGENTS_STARTUP.md §5` to document the per-task-class invocation guidance.

Also: surface the `bundle-parse5` requirement explicitly. Currently it's a hidden test-runner prerequisite — codifying it under `--install` (always run after install) closes that gap.

## Acceptance Criteria

- [ ] `bootstrapWorktree.mjs` accepts `--install` flag; idempotently runs `npm install` when `node_modules/` is absent in the worktree.
- [ ] `bootstrapWorktree.mjs` accepts `--build-all` flag; implies `--install`; runs `npm run build-all`.
- [ ] Both flags log progress + timing for observability (mirrors existing `copied: ...` style).
- [ ] `bundle-parse5` is run automatically as part of `--install` (or surfaced as a separate `--bundle-test-prereqs` flag — design decision deferred to implementation).
- [ ] `AGENTS_STARTUP.md §5 Worktree Bootstrap (Claude Code)` updated with per-task-class invocation guidance (which flag to use for backend vs frontend tickets).
- [ ] Empirical timing anchor cited inline in updated doc: ~17s for `--install` (this session), `--build-all` cost measured at implementation time.
- [ ] Skip when running from main checkout (existing pattern preserved).
- [ ] Existing tests still pass (no regression on the `--link-data` flow).
- [ ] New test or smoke-script verifying the install step is idempotent on re-invocation.

## Out of Scope

- **Replacing the symlink approach for `node_modules`.** Several harnesses (and several agents in this very session) considered symlinking `node_modules` from main checkout. The trade-off (canonical-path issues with some packages, git-worktree assumptions) doesn't justify it for MVP; full local install is more robust.
- **Build-flag matrix beyond `--install` / `--build-all`.** Granular flags like `--build-themes`, `--build-threads` are deferred unless empirical need surfaces.
- **Auto-detection of "do I need build-all" based on ticket scope.** Heuristic-based auto-bootstrap is a separate ergonomic concern; explicit flags are sufficient for the MVP.
- **Cross-harness applicability.** This ticket scope is Claude Code worktrees specifically (per `AGENTS_STARTUP.md §5`). Antigravity / Gemini CLI use shared-checkout patterns and don't need this script.

## Avoided Traps

- **Always-on install.** Rejected. Many tickets (docs-only, like PR #10348) don't need node_modules. Forcing an install on every bootstrap is friction without payoff.
- **Single mega-flag like `--full`.** Rejected. Naming explicit capabilities (`--install`, `--build-all`) lets agents pick the minimum-viable bootstrap for the task class.
- **Symlink node_modules from main checkout.** Rejected. Explicitly considered this session and denied at the harness security layer; even if permitted, canonical-path resolver semantics introduce risk for some packages.
- **Coupling with `--link-data`.** Rejected. `--link-data` is data-substrate concern (Memory Core SQLite); `--install` is dependency-substrate concern. Independent flags compose; coupling creates implicit-skip surprise.

## Related

- **`ai/scripts/bootstrapWorktree.mjs`** — script being extended.
- **`AGENTS_STARTUP.md §5`** — doc to update.
- **#10339** — empirical-anchor ticket where this gap surfaced (PR #10350 implementation hit the missing-`node_modules` blocker).
- **#10095** — original worktree-substrate ticket establishing the bootstrap pattern.
- **#10224** — `.neo-ai-data` symlink unification (companion `--link-data` flag established here).

Origin Session ID: `48197e2e-3e95-47eb-9eb8-bbb032948845`

Retrieval Hint: `query_raw_memories(query='bootstrapWorktree npm install build-all worktree dependencies dist parse5 node_modules')`

--- a/AGENTS_STARTUP.md
+++ b/AGENTS_STARTUP.md
@@ -76,9 +76,16 @@ Error [ERR_MODULE_NOT_FOUND]: Cannot find module '.../ai/mcp/server/github-workf
 ```bash
 node ai/scripts/bootstrapWorktree.mjs              # copy configs only
 node ai/scripts/bootstrapWorktree.mjs --link-data  # ALSO unify .neo-ai-data/ (recommended)
+node ai/scripts/bootstrapWorktree.mjs --link-data --install     # ALSO npm install + bundle-parse5
+node ai/scripts/bootstrapWorktree.mjs --link-data --build-all   # ALSO full Webpack build (frontend tickets)
 ```
 
 It copies the four `config.mjs` files from the main checkout (resolved via `git worktree list --porcelain`) and — with `--link-data` — symlinks the worktree's `.neo-ai-data/` to the main checkout's. Idempotent; no-op from the main checkout.
+
+**Per-task-class invocation guidance (per #10351):**
+- **Docs-only tickets** — `--link-data` is sufficient (no `node_modules` needed; the worktree filesystem itself + the data unification covers everything).
+- **Backend / MCP / unit-test tickets** — add `--install`. Empirical anchor: ~17s for `npm install` on a populated local cache (808 packages observed); `bundle-parse5` adds ~1-2s and IS required for the unit-test runner. Skips both if `node_modules/` already exists in the worktree.
+- **Frontend / Webpack-distribution tickets** — add `--build-all`. Implies `--install`; runs the full `npm run build-all` after dependencies are present. Use only when the ticket actually touches frontend bundles, themes, or Webpack thread distributions — backend tickets pay the full build cost for nothing otherwise.
 
 **Why `--link-data` matters (per #10224):** without it, each worktree gets its own empty `.neo-ai-data/sqlite/memory-core-graph.sqlite`, which means AgentIdentity nodes seeded in the main checkout are invisible to the worktree's MCP server. `bindAgentIdentity('neo-opus-4-7')` returns null, the mailbox throws `"no agent identity context bound"`, and A2A handshakes silently fail — the #10184 symptom from a different root cause than cache coherence. The symlink unifies the Memory Core substrate (SQLite + Chroma + concepts + backups) so ADR 0001's "one SQLite file shared across N processes" assumption holds across worktrees.
 

--- a/ai/scripts/bootstrapWorktree.mjs
+++ b/ai/scripts/bootstrapWorktree.mjs
@@ -220,22 +220,25 @@ export async function symlinkDataDir({mainCheckout, projectRoot, dir = '.neo-ai-
  */
 export async function installDependencies({projectRoot, log = console.log, exec = execFileAsync}) {
     const nodeModulesPath = path.join(projectRoot, 'node_modules');
+    let action;
 
     if (await exists(nodeModulesPath)) {
         log(`install skip (exists): node_modules`);
+        action = 'already-installed';
     } else {
         log(`installing dependencies (npm install)...`);
         const start = Date.now();
         await exec('npm', ['install'], {cwd: projectRoot});
         log(`installed dependencies in ${Math.round((Date.now() - start) / 1000)}s`);
+        action = 'installed';
     }
 
     log(`bundling parse5 (test-runner prerequisite)...`);
-    const start = Date.now();
+    const bundleStart = Date.now();
     await exec('npm', ['run', 'bundle-parse5'], {cwd: projectRoot});
-    log(`bundled parse5 in ${Math.round((Date.now() - start) / 1000)}s`);
+    log(`bundled parse5 in ${Math.round((Date.now() - bundleStart) / 1000)}s`);
 
-    return await exists(path.join(projectRoot, 'node_modules')) ? 'installed' : 'already-installed';
+    return action;
 }
 
 /**

--- a/ai/scripts/bootstrapWorktree.mjs
+++ b/ai/scripts/bootstrapWorktree.mjs
@@ -193,6 +193,78 @@ export async function symlinkDataDir({mainCheckout, projectRoot, dir = '.neo-ai-
     return 'linked';
 }
 
+/**
+ * @summary Installs the worktree's `node_modules` and bundles the parse5 test prerequisite.
+ *
+ * Worktrees off `origin/dev` start without `node_modules` (gitignored) AND without
+ * `dist/parse5.mjs` (gitignored test-runner prerequisite). Both are needed to run the
+ * Playwright unit-test suite or any SDK-consuming script. Adding these as opt-in flags
+ * here closes the manual-multi-step bootstrap gap surfaced empirically during #10339
+ * implementation (per #10351).
+ *
+ * Idempotent: skips `npm install` when `node_modules/` is already present (e.g.,
+ * symlinked from main, prior `--install` invocation, or manual `npm i`). Always runs
+ * `npm run bundle-parse5` because the bundle output lives under `dist/` (gitignored)
+ * and is cheap to rebuild — surfacing it explicitly under `--install` closes the
+ * "test runner fails with `Cannot find module '.../dist/parse5.mjs'`" friction.
+ *
+ * Cost anchor: ~17s for `npm install` on a populated local cache (808 packages,
+ * empirically observed during #10339 implementation, 2026-04-26). `bundle-parse5`
+ * adds ~1-2s. Friction-free when `node_modules` already exists (skip path is sub-millisecond).
+ *
+ * @param {object}   options
+ * @param {string}   options.projectRoot      Absolute path to the worktree root.
+ * @param {Function} [options.log]            Logger fn for action diagnostics.
+ * @param {Function} [options.exec]           execFile wrapper for dependency injection (testing).
+ * @returns {Promise<'already-installed'|'installed'>} Action taken.
+ */
+export async function installDependencies({projectRoot, log = console.log, exec = execFileAsync}) {
+    const nodeModulesPath = path.join(projectRoot, 'node_modules');
+
+    if (await exists(nodeModulesPath)) {
+        log(`install skip (exists): node_modules`);
+    } else {
+        log(`installing dependencies (npm install)...`);
+        const start = Date.now();
+        await exec('npm', ['install'], {cwd: projectRoot});
+        log(`installed dependencies in ${Math.round((Date.now() - start) / 1000)}s`);
+    }
+
+    log(`bundling parse5 (test-runner prerequisite)...`);
+    const start = Date.now();
+    await exec('npm', ['run', 'bundle-parse5'], {cwd: projectRoot});
+    log(`bundled parse5 in ${Math.round((Date.now() - start) / 1000)}s`);
+
+    return await exists(path.join(projectRoot, 'node_modules')) ? 'installed' : 'already-installed';
+}
+
+/**
+ * @summary Runs the full `npm run build-all` after ensuring dependencies are installed.
+ *
+ * Implies {@link installDependencies}. Required for tickets that touch the frontend
+ * Webpack distributions or themes — backend-only / MCP-only tickets do not need this.
+ *
+ * NOT idempotent in the same idempotent-skip sense as the other bootstrap helpers —
+ * `npm run build-all` re-runs every invocation. Webpack itself caches incremental builds,
+ * so re-runs against an already-built tree are still considerably faster than cold builds.
+ *
+ * @param {object}   options
+ * @param {string}   options.projectRoot      Absolute path to the worktree root.
+ * @param {Function} [options.log]            Logger fn for action diagnostics.
+ * @param {Function} [options.exec]           execFile wrapper for dependency injection (testing).
+ * @returns {Promise<'built'>} Action taken.
+ */
+export async function runBuildAll({projectRoot, log = console.log, exec = execFileAsync}) {
+    await installDependencies({projectRoot, log, exec});
+
+    log(`running full build (npm run build-all)...`);
+    const start = Date.now();
+    await exec('npm', ['run', 'build-all'], {cwd: projectRoot});
+    log(`build-all completed in ${Math.round((Date.now() - start) / 1000)}s`);
+
+    return 'built';
+}
+
 // -------------------------------------------------------------------------------------
 // CLI entry point. Runs only when invoked directly (node ai/scripts/bootstrapWorktree.mjs)
 // and not when imported by a test spec.
@@ -208,6 +280,8 @@ if (isMain) {
     const args     = new Set(process.argv.slice(2));
     const linkData = args.has('--link-data');
     const force    = args.has('--force');
+    const install  = args.has('--install');
+    const buildAll = args.has('--build-all');
 
     try {
         const mainCheckout = await resolveMainCheckout(projectRoot);
@@ -223,6 +297,14 @@ if (isMain) {
         if (linkData) {
             const symlinkResult = await symlinkDataDir({mainCheckout, projectRoot, force});
             console.log(`✓ Data symlink: ${symlinkResult}`);
+        }
+
+        if (buildAll) {
+            const buildResult = await runBuildAll({projectRoot});
+            console.log(`✓ Build: ${buildResult}`);
+        } else if (install) {
+            const installResult = await installDependencies({projectRoot});
+            console.log(`✓ Install: ${installResult}`);
         }
     } catch (e) {
         console.error('Bootstrap failed:', e.message);

--- a/test/playwright/unit/ai/scripts/bootstrapWorktree.spec.mjs
+++ b/test/playwright/unit/ai/scripts/bootstrapWorktree.spec.mjs
@@ -31,6 +31,8 @@ import path            from 'path';
 test.describe('ai/scripts/bootstrapWorktree', () => {
     let bootstrapWorktree;
     let symlinkDataDir;
+    let installDependencies;
+    let runBuildAll;
     let BOOTSTRAP_CONFIGS;
     let fakeMainCheckout;
     let fakeWorktree;
@@ -43,10 +45,12 @@ test.describe('ai/scripts/bootstrapWorktree', () => {
     ];
 
     test.beforeAll(async () => {
-        const mod         = await import('../../../../../ai/scripts/bootstrapWorktree.mjs');
-        bootstrapWorktree = mod.bootstrapWorktree;
-        symlinkDataDir    = mod.symlinkDataDir;
-        BOOTSTRAP_CONFIGS = mod.BOOTSTRAP_CONFIGS;
+        const mod           = await import('../../../../../ai/scripts/bootstrapWorktree.mjs');
+        bootstrapWorktree   = mod.bootstrapWorktree;
+        symlinkDataDir      = mod.symlinkDataDir;
+        installDependencies = mod.installDependencies;
+        runBuildAll         = mod.runBuildAll;
+        BOOTSTRAP_CONFIGS   = mod.BOOTSTRAP_CONFIGS;
     });
 
     test.beforeEach(async () => {
@@ -244,6 +248,124 @@ test.describe('ai/scripts/bootstrapWorktree', () => {
             const lstat = await fs.lstat(path.join(fakeMainCheckout, dataDir));
             expect(lstat.isDirectory()).toBe(true);
             expect(lstat.isSymbolicLink()).toBe(false);
+        });
+    });
+
+    // --------------------------------------------------------------------------------
+    // #10351 installDependencies + runBuildAll — opt-in flags that close the
+    // multi-step manual bootstrap gap (npm install + bundle-parse5 + optional build-all).
+    //
+    // These tests inject a mock `exec` to capture invoked commands without spawning real
+    // npm child processes. The pure-function form (with explicit dependency injection)
+    // mirrors the existing `bootstrapWorktree`/`symlinkDataDir` testability pattern.
+    // --------------------------------------------------------------------------------
+    test.describe('#10351 installDependencies', () => {
+        function makeMockExec() {
+            const calls = [];
+            const exec = async (cmd, args, opts) => {
+                calls.push({cmd, args, opts});
+                // Simulate npm install creating node_modules — flips the existence guard
+                // for downstream calls that re-check.
+                if (cmd === 'npm' && args[0] === 'install') {
+                    await fs.ensureDir(path.join(opts.cwd, 'node_modules'));
+                }
+                return {stdout: '', stderr: ''};
+            };
+            return {exec, calls};
+        }
+
+        test('runs `npm install` then `bundle-parse5` when node_modules is absent', async () => {
+            const {exec, calls} = makeMockExec();
+            const result = await installDependencies({
+                projectRoot: fakeWorktree,
+                exec,
+                log        : () => {}
+            });
+
+            expect(result).toBe('installed');
+            expect(calls).toHaveLength(2);
+            expect(calls[0].cmd).toBe('npm');
+            expect(calls[0].args).toEqual(['install']);
+            expect(calls[1].cmd).toBe('npm');
+            expect(calls[1].args).toEqual(['run', 'bundle-parse5']);
+
+            // Both invocations targeted the worktree (not main checkout).
+            for (const call of calls) {
+                expect(call.opts.cwd).toBe(fakeWorktree);
+            }
+        });
+
+        test('skips `npm install` when node_modules already exists, but always runs bundle-parse5', async () => {
+            // Pre-seed node_modules to simulate prior install / symlink.
+            await fs.ensureDir(path.join(fakeWorktree, 'node_modules'));
+
+            const {exec, calls} = makeMockExec();
+            const result = await installDependencies({
+                projectRoot: fakeWorktree,
+                exec,
+                log        : () => {}
+            });
+
+            expect(result).toBe('installed');
+            expect(calls).toHaveLength(1);
+            expect(calls[0].args).toEqual(['run', 'bundle-parse5']);
+        });
+
+        test('emits log lines for skip / install / bundle paths', async () => {
+            const logs   = [];
+            const {exec} = makeMockExec();
+
+            // First run — install branch
+            await installDependencies({projectRoot: fakeWorktree, exec, log: line => logs.push(line)});
+            expect(logs.join('\n')).toContain('installing dependencies');
+            expect(logs.join('\n')).toContain('bundling parse5');
+
+            // Second run — skip branch
+            const logs2 = [];
+            await installDependencies({projectRoot: fakeWorktree, exec, log: line => logs2.push(line)});
+            expect(logs2.join('\n')).toContain('install skip (exists)');
+            expect(logs2.join('\n')).toContain('bundling parse5');
+        });
+    });
+
+    test.describe('#10351 runBuildAll', () => {
+        function makeMockExec() {
+            const calls = [];
+            const exec = async (cmd, args, opts) => {
+                calls.push({cmd, args, opts});
+                if (cmd === 'npm' && args[0] === 'install') {
+                    await fs.ensureDir(path.join(opts.cwd, 'node_modules'));
+                }
+                return {stdout: '', stderr: ''};
+            };
+            return {exec, calls};
+        }
+
+        test('composes installDependencies then runs `npm run build-all`', async () => {
+            const {exec, calls} = makeMockExec();
+            const result = await runBuildAll({
+                projectRoot: fakeWorktree,
+                exec,
+                log        : () => {}
+            });
+
+            expect(result).toBe('built');
+            // Expected sequence: npm install, npm run bundle-parse5, npm run build-all
+            expect(calls).toHaveLength(3);
+            expect(calls[0].args).toEqual(['install']);
+            expect(calls[1].args).toEqual(['run', 'bundle-parse5']);
+            expect(calls[2].args).toEqual(['run', 'build-all']);
+        });
+
+        test('skips install when node_modules exists but still runs bundle-parse5 + build-all', async () => {
+            await fs.ensureDir(path.join(fakeWorktree, 'node_modules'));
+
+            const {exec, calls} = makeMockExec();
+            await runBuildAll({projectRoot: fakeWorktree, exec, log: () => {}});
+
+            expect(calls).toHaveLength(2);
+            expect(calls[0].args).toEqual(['run', 'bundle-parse5']);
+            expect(calls[1].args).toEqual(['run', 'build-all']);
         });
     });
 });

--- a/test/playwright/unit/ai/scripts/bootstrapWorktree.spec.mjs
+++ b/test/playwright/unit/ai/scripts/bootstrapWorktree.spec.mjs
@@ -306,7 +306,11 @@ test.describe('ai/scripts/bootstrapWorktree', () => {
                 log        : () => {}
             });
 
-            expect(result).toBe('installed');
+            // Return value MUST reflect the action actually taken (skip), not just the
+            // post-condition that node_modules exists. Per @neo-gemini-3-1-pro's PR #10352
+            // cycle 1 review: a returned-action contract that always reports `'installed'`
+            // because node_modules ends up present either way is semantically broken.
+            expect(result).toBe('already-installed');
             expect(calls).toHaveLength(1);
             expect(calls[0].args).toEqual(['run', 'bundle-parse5']);
         });

